In [38]:
# 데이터 불러오기
from sklearn.datasets import load_boston
# 데이터 분리
from sklearn.model_selection import train_test_split
# 선형회귀 모델
from sklearn.linear_model import LinearRegression
# 평균 제곱근 오차
from sklearn.metrics import mean_squared_error

import pandas as pd

In [3]:
boston_dataset = load_boston()

### 데이터 정보 Description

In [5]:
print(boston_dataset.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

### 데이터 구조 확인

In [7]:
boston_dataset.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

* 506개의 데이터
* 입력변수 13개
* 목표변수 : 가격 MEDV

In [9]:
boston_dataset.data.shape

(506, 13)

목표 변수
* 506개의 집 가격

In [10]:
boston_dataset.target # 목표변수

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
       35.4, 24.7, 31.6, 23.3, 19.6, 18.7, 16. , 22.2, 25. , 33. , 23.5,
       19.4, 22. , 17.4, 20.9, 24.2, 21.7, 22.8, 23.4, 24.1, 21.4, 20. ,
       20.8, 21.2, 20.3, 28. , 23.9, 24.8, 22.9, 23.9, 26.6, 22.5, 22.2,
       23.6, 28.7, 22.6, 22. , 22.9, 25. , 20.6, 28.4, 21.4, 38.7, 43.8,
       33.2, 27.5, 26.5, 18.6, 19.3, 20.1, 19.5, 19.5, 20.4, 19.8, 19.4,
       21.7, 22.8, 18.8, 18.7, 18.5, 18.3, 21.2, 19.2, 20.4, 19.3, 22. ,
       20.3, 20.5, 17.3, 18.8, 21.4, 15.7, 16.2, 18. , 14.3, 19.2, 19.6,
       23. , 18.4, 15.6, 18.1, 17.4, 17.1, 13.3, 17.8, 14. , 14.4, 13.4,
       15.6, 11.8, 13.8, 15.6, 14.6, 17.8, 15.4, 21

In [11]:
boston_dataset.target.shape

(506,)

In [13]:
x = pd.DataFrame(boston_dataset.data, columns = boston_dataset.feature_names)

### 데이터 프레임으로 담기

In [20]:
x

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

### 입력변수가 하나인 경우만 선택

#### 집의 나이를 기준으로 선형회귀 적용

In [40]:
x = x[['AGE']]

* 목표 변수 지정

In [41]:
y = pd.DataFrame(boston_dataset.target, columns=['MEDV'])

In [42]:
y

MEDV
0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
..    ...
501  22.4
502  20.6
503  23.9
504  22.0
505  11.9

[506 rows x 1 columns]

### 데이터 분리 (TR TS)
* 모델의 성능을 신빙성있게 평가하기 위해서 모델을 학습시키는 데이터와 평가하는 데이터를 나눈다.

* test_size = 전체 데이터 중 테스트데이터로 할당될 비율
* random_state = 테스트데이터 20%를 어떻게 고를지 정하는 파라미터 
* 값을 넘겨주지 않으면  실행할 때마다 매번 랜덤한 20% 를 테스트 셋으로 새롭게 고르게 됨

1. train_test_split은 4개의 값을 튜플로 리턴한다.
2. 4개의 리턴 값을 4개의 변수에 저장한다.
3. 각각 데이터프레임으로 이루어져있다.

In [43]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=5)

* train data set에는 404개, test data set에는 102개의 데이터가 들어있다.

In [44]:
print(x_train.shape, x_test.shape, y_train.shape, y_test.shape)

(404, 1) (102, 1) (404, 1) (102, 1)


#### x_train

In [45]:
x_train

AGE
33    95.0
283   24.8
418  100.0
502   76.7
402  100.0
..     ...
486   79.8
189   38.9
495   28.8
206   52.5
355   19.5

[404 rows x 1 columns]

#### x_test

In [46]:
x_test

AGE
226   86.5
292   23.4
90    66.1
373  100.0
273   51.8
..     ...
349   34.5
212   53.8
156   94.0
480   64.7
248   49.1

[102 rows x 1 columns]

#### y_train

In [47]:
y_train

MEDV
33   13.1
283  50.0
418   8.8
502  20.6
402  12.1
..    ...
486  19.1
189  34.9
495  23.1
206  24.4
355  20.6

[404 rows x 1 columns]

#### y_test

In [48]:
y_test

MEDV
226  37.6
292  27.9
90   22.6
373  13.8
273  35.2
..    ...
349  26.6
212  22.4
156  13.1
480  23.0
248  24.5

[102 rows x 1 columns]

### 모델 적용 및 학습

In [49]:
model = LinearRegression()

In [50]:
model.fit(x_train, y_train)

LinearRegression()

In [51]:
model.coef_

array([[-0.12402883]])

In [52]:
model.intercept_

array([31.04617413])

#### f(x) = model.intercept_(37.91248701) - model.coef_(-0.12402883)

### 예측값
* test set에 대한 예측값 구하기

In [53]:
y_test_prediction = model.predict(x_test)
y_test_prediction[:10] 

array([[20.31768041],
       [28.14389953],
       [22.84786852],
       [18.64329122],
       [24.62148078],
       [19.99520545],
       [19.75955068],
       [22.79825699],
       [20.92542167],
       [21.74401194]])

### 모델 평가
* test set의 실제 목표변수 값들이 들어있는 output인 y_test
* 예측값과 실제 output값을 비교해서 모델을 평가한다.
* 보통, 평균 제곱근 오차 RMSE 를 사용한다.

In [ ]:
mean_squared_error(y_test, y_test_prediction) # 평균 제곱 오차 까지만 구해짐

### 평균 제곱근 오차

In [54]:
mean_squared_error(y_test, y_test_prediction) ** 0.5 # 루트 사용 

8.236881612652455

* 이러한 모델로 집값을 예측한다면, 8천달러 정도의 오차가 있음을 짐작할 수 있다.

---

### 코드 정리

#### 동네의 범죄율, CRIM을 사용해서 선형 회귀 적용

In [ ]:
# 필요한 라이브러리 import
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

import pandas as pd  

# 보스턴 집 데이터 갖고 오기
boston_house_dataset = datasets.load_boston()

# 입력 변수를 사용하기 편하게 pandas dataframe으로 변환
X = pd.DataFrame(boston_house_dataset.data, columns=boston_house_dataset.feature_names)

# 목표 변수를 사용하기 편하게 pandas dataframe으로 변환
y = pd.DataFrame(boston_house_dataset.target, columns=['MEDV'])

X = X[['CRIM']]  # 범죄율 열만 사용

# train_test_split를 사용해서 주어진 데이터를 학습, 테스트 데이터로 나눈다
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=5)

linear_regression_model = LinearRegression()  # 선형 회귀 모델을 가지고 오고 
linear_regression_model.fit(X_train, y_train)  # 학습 데이터를 이용해서 모델을 학습 시킨다

y_test_predict = linear_regression_model.predict(X_test)  # 학습시킨 모델로 예측

# 평균 제곱 오차의 루트를 통해서 테스트 데이터에서의 모델 성능 판단
mse = mean_squared_error(y_test, y_test_predict)

mse ** 0.5